lib/Transforms/Instrumentation/ExtendedPointer.cc

```diff
@@ -0,0 +1,409 @@
+#define SEA_GET_FAT_SLOT0 "__sea_get_extptr_slot0_fp"
+#define SEA_SET_FAT_SLOT1 "__sea_set_extptr_slot1_fp"
+#define SEA_GET_FAT_SLOT1 "__sea_get_extptr_slot1_fp"
+#define SEA_COPY_FAT_SLOTS "__sea_copy_extptr_slots_fp"
+#define SEA_RECOVER_FAT_PTR "__sea_recover_pointer_fp"
+#else
+#define SEA_SET_FAT_SLOT0 "__sea_set_extptr_slot0_hm"
+#define SEA_GET_FAT_SLOT0 "__sea_get_extptr_slot0_hm"
+#define SEA_SET_FAT_SLOT1 "__sea_set_extptr_slot1_hm"
+#define SEA_GET_FAT_SLOT1 "__sea_get_extptr_slot1_hm"
+#define SEA_COPY_FAT_SLOTS "__sea_copy_extptr_slots_hm"
+#define SEA_RECOVER_FAT_PTR "__sea_recover_pointer_hm"
+#endif
+
+//static cl::opt<bool> SingleErrorBB("bounds-checking-single-trap",
+//                                  cl::desc("Use one trap block per function"));
+
+STATISTIC(ChecksAdded, "Bounds checks added");
+STATISTIC(ChecksSkipped, "Bounds checks skipped");
+STATISTIC(ChecksUnable, "Bounds checks unable to add");
+
+typedef IRBuilder<TargetFolder> BuilderTy;
+
+namespace {
+struct ExtendedPointer : public FunctionPass {
+  static char ID;
+  ExtendedPointer() : FunctionPass(ID){}
+
+  bool runOnFunction(Function &F) override;
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.addRequired<TargetLibraryInfoWrapperPass>();
+  }
+
+private:
+  const TargetLibraryInfo *TLI;
+  ObjectSizeOffsetEvaluator *ObjSizeEval;
+  BuilderTy *Builder;
+  Instruction *Inst;
+  BasicBlock *ErrorBB;
+  Type *IntptrTy;
+  LLVMContext *Ctx;
+  Module *Mod;
+  Function* m_setFatSlot0; // set ptr base
+  Function* m_setFatSlot1; // set ptr size
+  Function* m_getFatSlot0; // get ptr base
+  Function* m_getFatSlot1; // get ptr size
+  Function* m_copyFatSlots; // copy ptr info
+  Function* m_recoverFatPtr; // NEW: for ptr embed method only
+  Value* OrigPtr;
+
+
+  BasicBlock *getErrorBB();
+  void emitBranchToTrap(Value *Cmp = nullptr);
+  bool instrument(Value *Ptr, Value *Val, const DataLayout &DL);
+  bool instrumentAddress(Value *Ptr, const DataLayout &DL, Value *BasePtr = nullptr);
+  // Value* memToShadow(Value *mem, BuilderTy *builder);
+};
+}
+
+char ExtendedPointer::ID = 0;
+
+/// getErrorBB - create a basic block that traps. All overflowing conditions
+/// branch to this block. There's only one trap block per function.
+BasicBlock *ExtendedPointer::getErrorBB() {
+  if (ErrorBB)
+    return ErrorBB;
+
+  Function *Fn = Inst->getParent()->getParent();
+  Module *Md = Fn->getParent();
+  LLVMContext &ctx = Md->getContext();
+  IRBuilder<>::InsertPointGuard Guard(*Builder);
+  ErrorBB = BasicBlock::Create(Fn->getContext(), "bound_overflow", Fn);
+  Builder->SetInsertPoint(ErrorBB);
+
+//  llvm::Value *F = Intrinsic::getDeclaration(Fn->getParent(), Intrinsic::trap);
+  AttrBuilder AB;
+  AB.addAttribute(Attribute::NoReturn);
+  AttributeList as = AttributeList::get(ctx, AttributeList::FunctionIndex, AB);
+  auto errorFn = dyn_cast<Function>(
+      Md->getOrInsertFunction("__VERIFIER_error", as, Type::getVoidTy(ctx)));
```

Comment:
Add comment: This generates a fn in the BC that is impl in seahorn (found at linked time)

Reply:
I think using `verifier.error` should be better. `__VERIFIER_error()` is a function defined in SV-COMP. We have a pass to promote it to `verifier.error` that we use internally to ensure that there are no name clashes between our special functions and user defined ones. 

There should be a standard utility to access the standard functions. This is something that is currently missing with raw names being used all over the place. 

@danblitzhou did using `verifier.error` cause a problem? Maybe I am missing something.

Reply:
I am not sure if using verifier.error would result in different behaviour or not, you can try swapping

Reply:
Using `__VERIFIER_error` is currently preferred since there are no API to get a unifrom declaration of the internal intrinsic `verifier.error()`.